python/ray/train/v2/tests/test_metrics.py

```diff
@@ -54,42 +131,81 @@ def test_worker_metrics_callback(monkeypatch):
         "get_train_context",
         lambda: mock_train_context,
     )
-    monkeypatch.setattr(ray.train.v2._internal.callbacks.metrics, "Gauge", MockGauge)
 
     callback = WorkerMetricsCallback(
         train_run_context=TrainRunContext(run_config=RunConfig(name="test_run_name"))
     )
     callback.after_init_train_context()
 
     # Check if the gauges is updated with the correct metrics
-    callback.on_report(1.0)
-    time.sleep(0.1)
-    assert callback._metrics_gauges["train_report_total_blocked_time_s"].get() == 1.0
+    with callback.on_report():
+        pass
+    assert (
+        callback._metrics[WorkerMetrics.REPORT_TOTAL_BLOCKED_TIME_S].get_value() == 1.0
+    )
 
     # Check if the gauges is updated with the correct metrics
-    callback.on_report(1.0)
-    time.sleep(0.1)
-    assert callback._metrics_gauges["train_report_total_blocked_time_s"].get() == 2.0
+    with callback.on_report():
+        pass
+    assert (
+        callback._metrics[WorkerMetrics.REPORT_TOTAL_BLOCKED_TIME_S].get_value() == 3.0
+    )
 
 
 def test_controller_metrics_callback(monkeypatch):
+    monkeypatch.setattr(ray.train.v2._internal.metrics.base, "Gauge", MockGauge)
+    mock_time_monotonic(monkeypatch, [0.0, 1.0, 10.0, 12.0])
```

Comment:
also test shutdown logic with before_controller_shutdown and before_shutdown for worker and controller metrics callback